src/debug.cc

```diff
@@ -91,6 +416,209 @@ ResyncDebugLog(FILE *newFile)
     TheLog.file_ = newFile;
 }
 
+/* DebugChannel */
+
+void
+DebugChannel::stopEarlyMessageCollection()
+{
+    logSaved();
+}
+
+void
+Debug::ForgetSaved()
+{
+    auto &module = Module();
+    (void)module.cacheLogChannel.releaseEarlyMessages();
+    (void)module.stderrChannel.releaseEarlyMessages();
+    (void)module.syslogChannel.releaseEarlyMessages();
+}
+
+void
+Debug::PrepareToDie()
+{
+    Module().prepareToDie();
+}
+
+void
+DebugChannel::logSavedAnd(DebugChannel *theirChannelOrNil)
+{
+    const LoggingSectionGuard sectionGuard;
+
+    assert(this != theirChannelOrNil);
+    const auto ours = releaseEarlyMessages();
+    const auto theirs = theirChannelOrNil ? theirChannelOrNil->releaseEarlyMessages() : nullptr;
+
+    size_t ourPos = 0;
+    size_t theirPos = 0;
+    size_t ourCount = ours ? ours->size() : 0;
+    size_t theirCount = theirs ? theirs->size() : 0;
+
+    const auto writtenEarlier = written;
+    while (ourPos < ourCount || theirPos < theirCount) {
+        // write in the order of increasing header.recordNumber;
+        // log() will filter out any duplicates
+        const auto &message = (ourPos < ourCount &&
+            !(theirPos < theirCount && (*theirs)[theirPos].header.recordNumber < (*ours)[ourPos].header.recordNumber)) ?
+            (*ours)[ourPos++] : (*theirs)[theirPos++];
+        if (Debug::Enabled(message.header.section, message.header.level))
+            log(message.header, message.body);
+    }
```

Comment:
>  replace back_inserter() with something that will call log() for Enabled() messages

I think we can do this by means of a helper class with back_insert_iterator's semantics which is passed to the merge algorithm instead of the back inserter. I have created a separate test to check this suggestion. For example:
```
class LogInserter : public std::iterator<std::forward_iterator_tag, DebugMessage>
{
public:
    LogInserter(DebugChannel &ch) : channel(ch) {}
    LogInserter &operator=(const DebugMessage &message) {
        if (Debug::Enabled(message.header.section, message.header.level))
            channel.log(message.header, message.body);
    }
    LogInserter &operator*() { return *this; }
    LogInserter &operator++() { return *this; }
private:
    DebugChannel &channel;
};
```

Reply:
No way to do that with a lambda, right?

Reply:
We need something that acts like back_insert_iterator - I haven't got an idea how to do this with lambda.

Reply:
Let's use an iterator, with the following adjustments:

0. I suggest calling the new iterator DebugMessageLogger or similar. If it is going to be declared inside the  DebugChannel class, then we can call it simply Logger.
1. Avoid using deprecated-in-C++17 std::iterator. This article might help: https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/
2. If possible, define just an _output_ iterator instead of a more powerful forward iterator. AFAICT, we just need to satisfy the LegacyOutputIterator requirements here. This may help simplify the code despite item 1 above.
3. Please check whether it makes sense to reuse the new output iterator elsewhere in debug.cc.